src/config/config_generator.cc

```diff
@@ -0,0 +1,449 @@
+/*GRB*
+
+Gerbera - https://gerbera.io/
+
+    config_generator.cc - this file is part of Gerbera.
+
+    Copyright (C) 2016-2018 Gerbera Contributors
+
+    Gerbera is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License version 2
+    as published by the Free Software Foundation.
+
+    Gerbera is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Gerbera.  If not, see <http://www.gnu.org/licenses/>.
+
+    $Id$
+*/
+
+/// \file config_generator.cc
+
+
+#include <string>
+#include <mxml/mxml.h>
+#include <common.h>
+#include <tools.h>
+#include <metadata_handler.h>
+#include <uuid/uuid.h>
+#include "config_generator.h"
+
+using namespace zmm;
+using namespace mxml;
+
+ConfigGenerator::ConfigGenerator() {}
+
+ConfigGenerator::~ConfigGenerator() {}
+
+std::string ConfigGenerator::generate(std::string userHome, std::string configDir, std::string prefixDir, std::string magicFile) {
+  Ref<Element> config(new Element(_("config")));
+  config->setAttribute(_("version"), String::from(CONFIG_XML_VERSION));
+  config->setAttribute(_("xmlns"), _(XML_XMLNS) + CONFIG_XML_VERSION);
+  config->setAttribute(_("xmlns:xsi"), _(XML_XMLNS_XSI));
+  config->setAttribute(_("xsi:schemaLocation"),
+                       _(XML_XMLNS) + CONFIG_XML_VERSION + " " + XML_XMLNS + CONFIG_XML_VERSION + ".xsd");
+
+  Ref<Comment> docinfo(new Comment(_("\n\
+     See http://gerbera.io or read the docs for more\n\
+     information on creating and using config.xml configration files.\n\
+    "),true));
+
+  config->appendChild(RefCast(docinfo, Node));
+
+  config->appendElementChild(generateServer(userHome, configDir, prefixDir));
+
+  config->appendElementChild(generateImport(prefixDir, magicFile));
+
+  config->appendElementChild(generateTranscoding());
+
+  config->indent();
+  return std::string(config->print().c_str());
+}
+
+Ref<Element> ConfigGenerator::generateServer(std::string userHome, std::string configDir, std::string prefixDir) {
+  Ref<Element> server(new Element(_("server")));
+
+  server->appendElementChild(generateUi());
+  server->appendTextChild(_("name"), _(PACKAGE_NAME));
+
+  server->appendElementChild(generateUdn());
+
+  std::string homepath = userHome + DIR_SEPARATOR + configDir;
+  server->appendTextChild(_("home"), String::copy(homepath.c_str()));
```

Comment:
This line had a memory leak.  By passing `std::string` into the function, but then returning the `Ref<Element>`.  The  **std::string** value is not used until later in the **generate** method.

I chose the `Ref<String>` implementation of `String::copy` which duplicates the string.  I previously tried using `strdup(...)` but this creates a `char*` which needs to be **free'd** later.

Any insight or suggestions here would be appreciated.